### DISCUSSION TITLE
How to prevent all period tasks executed immediately when beat restart?

### DISCUSSION THREAD

**Discussion by tsingwang** [celery/celery#8076] [Q&A] [answered]
Restarting beat will perform all tasks immediately. Even if it's not the scheduling time in the config, it will still execute, which is not the behavior I expect. I expect it not to be executed on startup, how to do, thanks.
```
celery -A project worker -B -l info
```
```
CELERY_BEAT_SCHEDULE = {
    'sync_contacts': {
        'task': 'apps.portal.tasks.sync_contacts',
        'schedule': crontab(minute=0, hour=0),
    },
}
```
```
[2023-02-20 11:09:56,661: INFO/Beat] beat: Starting...
[2023-02-20 11:09:56,667: INFO/MainProcess] Connected to redis://localhost:6379/0
[2023-02-20 11:09:56,672: INFO/MainProcess] mingle: searching for neighbors
[2023-02-20 11:09:56,722: INFO/Beat] Scheduler: Sending due task sync_contacts (apps.portal.tasks.sync_contacts)
```

**Comment by norbertcyran** [comment, marked as answer]
As far as I see, it's a local environment, so my guess is that celery beat isn't turned on all the time. Beat keeps track of timestamps when periodic tasks were executed and based on those timestamps next run times are calculated. So basically, when the beat is down during the scheduled time (in your case, at midnight), the task obviously will not execute, and at the time when you restart the beat, it will notice that the task is due since midnight and it will execute the task right away. When the periodic task is not yet written to the database, the last run defaults to the current timestamp, so the task will **not** be executed.

In the production environment, celery beat usually will be up all the time, and short restarts will not cause all periodic tasks to be executed, so probably you don't have to bother about that.

If you really want to disable that behavior, you can try out adding `--scheduler celery.beat.Scheduler` argument, but I wouldn't really recommend doing that on production.

By the way, running beat by `-B` argument to the worker is also not recommended and on production, beat should be started by `celery -A project beat`

More info: https://docs.celeryq.dev/en/stable/userguide/periodic-tasks.html#starting-the-scheduler

**Reply by tsingwang** [reply]
Thanks for your reply.

But this is not my scenario though, the beat service is just doing an upgrade and only down for a few seconds. For example, if that period task has been executed once today and restart beat later, then all tasks will be scheduled to execute.

But I tested it in my local and found that it won't be scheduled again if restarted for a short time, I guess it may be related to my configuration `CELERY_RESULT_EXPIRES = 60 * 60`, it seems that it will be rescheduled after restarting after more than an hour. I haven't found any information to support this, so I need to experiment again.

I know your production recommendations, thanks.

**Reply by tsingwang** [reply]
After a day of testing, there is still no solution, no related with the `CELERY_RESULT_EXPIRES`, no related with the `celerybeat-schedule` file.

**Reply by norbertcyran** [reply]
What celery version and what broker do you use? Do you use `acks_late`/`eta`/`countdown`? Do all periodic tasks get executed after an hour of inactivity or only a part of them?

Did you try that `--scheduler celery.beat.Scheduler` workaround?

**Reply by tsingwang** [reply]
Wow! Use `--scheduler celery.beat.Scheduler` solved my problem. You really helped me, thanks.

Celery use `PersistentScheduler` default, I checked and parsed `celerybeat-schedule` file before, nothing was found.

Look at the code cursorily,  `setup_schedule` in `PersistentScheduler` do a `sync`.

**Comment by erw-emd-dk** [comment]
I had the same problem as OP, but solved it by setting `CELERY_BEAT_SYNC_EVERY = 1`. I don't understand exactly why, but it made sure that my task only ran at the time specified in my `crontab()` schedule, and not at beat startup.

https://docs.celeryq.dev/en/stable/userguide/configuration.html#beat-sync-every
